src/targets/node/watchdogSpawn.ts

```diff
@@ -0,0 +1,61 @@
+import { spawn } from 'child_process';
+import { join } from 'path';
+
+export interface IWatchdogInfo {
+  /**
+   * Observed process ID.
+   */
+  pid?: string;
+
+  /**
+   * If set to true, this indicates that the process the watchdog is monitoring
+   * was not started with the bootloader. In order to debug it, we must tell
+   * CDP to force it into debugging mode manually.
+   */
+  dynamicAttach?: boolean;
+
+  /**
+   * Process script name, for cosmetic purposes.
+   */
+  scriptName: string;
+
+  /**
+   * URL of the inspector running on the process.
+   */
+  inspectorURL: string;
+
+  /**
+   * Address on the debugging server to attach to.
+   */
+  ipcAddress: string;
+
+  /**
+   * Whether the process is waiting for the debugger to attach.
+   */
+  waitForDebugger: boolean;
+
+  /**
+   * Parent process ID.
+   */
+  ppid?: string;
+}
+
+const watchdogPath = join(__dirname, 'watchdog.js');
+
+/**
+ * Spawns a watchdog attached to the given process.
+ */
+export function spawnWatchdog(execPath: string, { ipcAddress, ...inspectorInfo }: IWatchdogInfo) {
```

Comment:
Can you explain what the watchDog is? Why does it run in a separate process?

Reply:
Pavel could explain more, the idea of the watchdog is to give us a target API for Node.js. Out of the box, Node doesn't include the Target namespace--there's no representation of child processes within its inspector protocol. The watchdog sits as a proxy which connects the Node process' CDP to the server we stand up within the BaseNodeLauncher. It handles the target IDs, opener IDs, making the node process 'look' like a webworker or any other thread within CDP.

Reply:
Oh, I remember that now.